handler/src/main/java/io/netty/handler/ssl/OpenSslKeyMaterialManager.java

```diff
@@ -79,101 +74,61 @@ void setKeyMaterial(ReferenceCountedOpenSslEngine engine) throws SSLException {
-            }
-            CertificateRequestedCallback.KeyMaterial material = new CertificateRequestedCallback.KeyMaterial(
-                    certChain, pkey);
-
-            // Reset to 0 so we do not free these. This is needed as the client certificate callback takes ownership
-            // of both the key and the certificate if they are returned from this method, and thus must not
-            // be freed here.
-            certChain = pkey = 0;
-            return material;
+            keyMaterial = provider.chooseKeyMaterial(engine.alloc, alias);
+            SSL.setKeyMaterialClientSide(
+                    engine.sslPointer(), certOut, keyOut, keyMaterial.chain(), keyMaterial.privateKey());
         } catch (SSLException e) {
             throw e;
         } catch (Exception e) {
             throw new SSLException(e);
         } finally {
-            freeBio(keyBio);
-            freeBio(keyCertChainBio);
-            SSL.freePrivateKey(pkey);
-            SSL.freeX509Chain(certChain);
+            if (keyMaterial != null) {
+                keyMaterial.release();
+            }
         }
     }
 
-    private void setKeyMaterial(long ssl, String alias, ByteBufAllocator allocator) throws SSLException {
-        long keyBio = 0;
-        long keyCertChainBio = 0;
-        long keyCertChainBio2 = 0;
-
-        try {
-            // TODO: Should we cache these and so not need to do a memory copy all the time ?
-            X509Certificate[] certificates = keyManager.getCertificateChain(alias);
-            if (certificates == null || certificates.length == 0) {
-                return;
-            }
-
-            PrivateKey key = keyManager.getPrivateKey(alias);
-
-            // Only encode one time
-            PemEncoded encoded = PemX509Certificate.toPEM(allocator, true, certificates);
-            try {
-                keyCertChainBio = toBIO(allocator, encoded.retain());
-                keyCertChainBio2 = toBIO(allocator, encoded.retain());
-
-                if (key != null) {
-                    keyBio = toBIO(allocator, key);
-                }
-                SSL.setCertificateBio(ssl, keyCertChainBio, keyBio, password);
-
-                // We may have more then one cert in the chain so add all of them now.
-                SSL.setCertificateChainBio(ssl, keyCertChainBio2, true);
-            } finally {
-                encoded.release();
-            }
-        } catch (SSLException e) {
-            throw e;
-        } catch (Exception e) {
-            throw new SSLException(e);
-        } finally {
-            freeBio(keyBio);
-            freeBio(keyCertChainBio);
-            freeBio(keyCertChainBio2);
+    private String chooseClientAlias(ReferenceCountedOpenSslEngine engine,
+                                       String[] keyTypes, X500Principal[] issuer) {
+        X509KeyManager manager = provider.keyManager();
+        if (manager instanceof X509ExtendedKeyManager) {
```

Comment:
did you consider have a class that extends `OpenSslKeyMaterialManager`, overrides these methods, and requires `X509ExtendedKeyManager` to avoid casting here (and below)?

Reply:
Yeah that said I thought I would just keep it simple for now. Its not clear to me if this would give any gain at all.